### PR TITLE
Fix promote-fn to work with named anonymous fns

### PR DESCRIPTION
**Pull request by luxbock**
- If the fn is named then use that name without asking

### CODE REVIEW

**Comment by expez** [comment]
This is great, thanks!

Would you mind adding a test or two demonstrating/validating the new behavior?

**Comment by luxbock** [comment]
My first time writing tests for Elisp :)

Right now promote-fn doesn't work if it's called from the hash-tag or before the argument vector of an anonymous function, so I think I'll try to fix that as well.

**Comment by expez** [comment]
> Right now promote-fn doesn't work if it's called from the hash-tag or before the argument vector of an anonymous function, so I think I'll try to fix that as well.

Cool!

**Comment by luxbock** [comment]
Hmm think this might still need a bit more work.

**Comment by luxbock** [comment]
I believe this should work now.

**Comment by luxbock** [comment]
Except Travis is failing. I don't get it. The tests pass when I run them locally.

**Comment by benedekfazekas** [comment]
let me rerun the build. not as if i don't trust travis...

**Comment by benedekfazekas** [comment]
nope. it hangs with 24.3. try to run the tests with 24.3

**Comment by luxbock** [comment]
Grr... The problem was that `<` only works with two arguments prior to 24.1. Why have Polish notation at all if `<` can't accept variable number of args!?

File: clj-refactor.el [hunk 1]
```diff
@@ -429,17 +429,31 @@ e.g. `re-search-forward'"
...
-      (error "Can't find definition of anonymous function!"))))
+  (if (zerop (first (paredit-current-parse-state)))
+      (error "Not inside a s-expression.")
+    (let* ((pt-orig (point))
+           (search-bound (cljr--point-after 'cljr--goto-toplevel))
+           found-fn-p)
+      (while (not found-fn-p)
+        (paredit-backward-up)
+        (-if-let (fn-beg (cljr--point-for-anon-function))
+            (let ((fn-end (save-excursion (paredit-forward) (point))))
```
Comment (expez) [inline clj-refactor.el]: This should probably be `(cljr--point-after 'paredit-forward)`

File: clj-refactor.el [hunk 2]
```diff
@@ -429,17 +429,31 @@ e.g. `re-search-forward'"
...
-        (progn (goto-char (max (or literal (point-min)) (or fn (point-min))))
-               (when (looking-back "#")
-                 (backward-char)))
-      (error "Can't find definition of anonymous function!"))))
+  (if (zerop (first (paredit-current-parse-state)))
+      (error "Not inside a s-expression.")
+    (let* ((pt-orig (point))
+           (search-bound (cljr--point-after 'cljr--goto-toplevel))
+           found-fn-p)
+      (while (not found-fn-p)
```
Comment (expez) [inline clj-refactor.el]: Won't this loop forever if there's no fn definition between here and the toplevel of the file?  I think maybe this needs to be guarded with a check on `cljr--toplevel-p?`

Comment (luxbock) [inline clj-refactor.el]: It throws via `(error "Can't find definition of anonymous function!")` if point reaches the top-level, i.e. `search-bound`, which in turn is found via calling `cljr--goto-toplevel`, which also uses `paredit-backward-up`, so I'm fairly sure it should work.

**Comment by expez** [comment]
Thank for addressing these issues!

**Pull request merged**